lib/logthrsource/logthrfetcherdrv.c

```diff
@@ -0,0 +1,277 @@
+ *
+ * As an additional exemption you are allowed to compile & link against the
+ * OpenSSL libraries as published by the OpenSSL project. See the file
+ * COPYING for details.
+ *
+ */
+
+#include "logthrfetcherdrv.h"
+#include "messages.h"
+
+static EVTTAG *
+evt_tag_driver(LogThreadedFetcherDriver *f)
+{
+  return evt_tag_str("driver", f->super.super.super.id);
+}
+
+static inline void
+_thread_init(LogThreadedFetcherDriver *self)
+{
+  msg_trace("Fetcher thread_init()", evt_tag_driver(self));
+  if (self->thread_init)
+    self->thread_init(self);
+}
+
+static inline void
+_thread_deinit(LogThreadedFetcherDriver *self)
+{
+  msg_trace("Fetcher thread_deinit()", evt_tag_driver(self));
+  if (self->thread_deinit)
+    self->thread_deinit(self);
+}
+
+static inline gboolean
+_connect(LogThreadedFetcherDriver *self)
+{
+  msg_trace("Fetcher connect()", evt_tag_driver(self));
+  if (!self->connect)
+    return TRUE;
+
+  if (!self->connect(self))
+    {
+      msg_debug("Error establishing connection", evt_tag_driver(self));
+      return FALSE;
+    }
+
+  return TRUE;
+}
+
+static inline void
+_disconnect(LogThreadedFetcherDriver *self)
+{
+  msg_trace("Fetcher disconnect()", evt_tag_driver(self));
+  if (self->disconnect)
+    self->disconnect(self);
+}
+
+static void
+_start_reconnect_timer(LogThreadedFetcherDriver *self)
+{
+  iv_validate_now();
+  self->reconnect_timer.expires  = iv_now;
+  self->reconnect_timer.expires.tv_sec += self->time_reopen;
+  iv_timer_register(&self->reconnect_timer);
+}
+
+static void
+_worker_run(LogThreadedSourceDriver *s)
+{
+  LogThreadedFetcherDriver *self = (LogThreadedFetcherDriver *) s;
+
+  iv_init();
+
+  iv_event_register(&self->wakeup_event);
+  iv_event_register(&self->shutdown_event);
+
+  _thread_init(self);
+  if (_connect(self))
+    iv_task_register(&self->fetch_task);
+  else
+    _start_reconnect_timer(self);
+
+  iv_main();
+
+  _disconnect(self);
+  _thread_deinit(self);
+
+  iv_deinit();
+}
+
+static void
+_worker_request_exit(LogThreadedSourceDriver *s)
+{
+  LogThreadedFetcherDriver *self = (LogThreadedFetcherDriver *) s;
+
+  self->under_termination = TRUE;
+
+  if (self->request_exit)
+    self->request_exit(self);
+
+  iv_event_post(&self->shutdown_event);
+}
+
+static void
+_wakeup(LogThreadedSourceDriver *s)
+{
+  LogThreadedFetcherDriver *self = (LogThreadedFetcherDriver *) s;
+
+  if (!self->under_termination)
+    iv_event_post(&self->wakeup_event);
+}
+
+static inline void
+_schedule_next_fetch_if_free_to_send(LogThreadedFetcherDriver *self)
+{
+  if (log_threaded_source_free_to_send(&self->super))
+    iv_task_register(&self->fetch_task);
+}
+
+static void
+_fetch(gpointer data)
+{
+  LogThreadedFetcherDriver *self = (LogThreadedFetcherDriver *) data;
+
+  msg_trace("Fetcher fetch()", evt_tag_driver(self));
+
+  LogThreadedFetchResult fetch_result = self->fetch(self);
+
+  switch (fetch_result.result)
+    {
+    case THREADED_FETCH_ERROR:
```

Comment:
"nitpick" : factor out the case branches into small `static on_xxx` functions.

Reply:
Done.